plugins/modules/virt_volume.py

```diff
@@ -0,0 +1,711 @@
+        elif state in ['undefined', 'absent']:
+            entries = v.list_volumes()
+            if name in entries:
+                res['changed'] = True
+                res['msg'] = v.delete(name)
+        elif state in ['deleted']:
+            entries = v.list_volumes()
+            if name in entries:
+                v.wipe(name, mode)
+                res['changed'] = True
+                res['msg'] = v.delete(name)
+        else:
+            module.fail_json(msg="unexpected state")
+
+        return VIRT_SUCCESS, res
+
+    if command:
+        if command in ENTRY_COMMANDS:
+            if not name:
+                module.fail_json(msg="%s requires 1 argument: name" % command)
+            if command == 'define':
+                if not xml:
+                    module.fail_json(msg="define requires xml argument")
+                try:
+                    v.get_pool(name)
+                except EntryNotFound:
+                    v.define(name, xml)
+                    res = {'changed': True, 'created': name}
+                return VIRT_SUCCESS, res
+            elif command == 'build':
+                res = v.build(name, mode)
+                if not isinstance(res, dict):
+                    res = {'changed': True, command: res}
+                return VIRT_SUCCESS, res
+            elif command == 'delete':
+                res = v.delete(name, mode)
+                if not isinstance(res, dict):
+                    res = {'changed': True, command: res}
+                return VIRT_SUCCESS, res
+            res = getattr(v, command)(name)
+            if not isinstance(res, dict):
+                res = {command: res}
+            return VIRT_SUCCESS, res
+
+        elif hasattr(v, command):
+            res = getattr(v, command)()
+            if not isinstance(res, dict):
+                res = {command: res}
+            return VIRT_SUCCESS, res
+
+        else:
+            module.fail_json(msg="Command %s not recognized" % command)
+
+    module.fail_json(msg="expected state or command parameter to be specified")
+
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(aliases=['volume']),
+            pool=dict(required=True),
+            state=dict(choices=['present', 'absent', 'undefined', 'deleted']),
+            command=dict(choices=ALL_COMMANDS),
+            uri=dict(default='qemu:///system'),
+            xml=dict(),
+            mode=dict(choices=ALL_MODES),
+        ),
+        supports_check_mode=True
+    )
+
+    if not HAS_VIRT:
+        module.fail_json(
+            msg='The `libvirt` module is not importable. Check the requirements.'
+        )
+
+    if not HAS_XML:
+        module.fail_json(
+            msg='The `lxml` module is not importable. Check the requirements.'
+        )
+
+    rc = VIRT_SUCCESS
+    try:
+        rc, result = core(module)
+    except Exception as e:
```

Comment:
For now, it's probably enough to use `EntryNotFound`and since Volumes are also entries. and then define more specific Exceptions during the refactoring maybe?